pkg/distributor/distributor.go

```diff
@@ -432,13 +439,36 @@ func (d *Distributor) shardStream(stream logproto.Stream, streamSize int, userID
 		level.Info(logger).Log("msg", "sharding request", "shard_count", shardCount)
 	}
 
-	streamLabels := labelTemplate(stream.Labels)
-	streamPattern := streamLabels.String()
+	if shardCount > len(stream.Entries) {
+		shardsOrder := d.randomizeShardsOrder(shardCount)
+		return d.divideEntriesBetweenShards(logger, userID, len(stream.Entries), shardStreamsCfg, stream, shardsOrder)
+	}
+
+	return d.divideEntriesBetweenShards(logger, userID, shardCount, shardStreamsCfg, stream, nil)
+}
+
+func (d *Distributor) randomizeShardsOrder(shardCount int) []int {
+	shardsOrder := make([]int, shardCount)
+	for i := 0; i < shardCount; i++ {
+		shardsOrder = append(shardsOrder, i)
+	}
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(shardsOrder), func(i, j int) { shardsOrder[i], shardsOrder[j] = shardsOrder[j], shardsOrder[i] })
+	return shardsOrder
+}
 
+func (d *Distributor) divideEntriesBetweenShards(logger log.Logger, userID string, shardCount int, shardStreamsCfg *shardstreams.Config, stream logproto.Stream, alternateShardsOrder []int) ([]uint32, []streamTracker) {
 	derivedKeys := make([]uint32, 0, shardCount)
 	derivedStreams := make([]streamTracker, 0, shardCount)
+	streamLabels := labelTemplate(stream.Labels)
+	streamPattern := streamLabels.String()
+
 	for i := 0; i < shardCount; i++ {
-		shard, ok := d.createShard(shardStreamsCfg, stream, streamLabels, streamPattern, shardCount, i)
+		j := i
+		if alternateShardsOrder != nil {
```

Comment:
It's preferred to check `len(alternateShardsOrder) > 0` because it works for a 0-length slice and a nil slice while `alternateShardsOrder != nil` only works for a nil slice.

Reply:
good call, pushed.